### PR TITLE
*: Introduce schema version

### PR DESCRIPTION
**Pull request by ngaut**
Make schema change to an atomic operation.

### CODE REVIEW

File: ddl/ddl.go [hunk 1]
```diff
@@ -82,27 +87,54 @@ func (d *ddl) CreateSchema(ctx context.Context, schema model.CIStr) (err error)
...
 	}
 	info := &model.DBInfo{Name: schema}
 	info.ID, err = meta.GenGlobalID(d.store)
 	if err != nil {
 		return errors.Trace(err)
 	}
-	err = d.writeSchemaInfo(info)
+
+	err = kv.RunInNewTxn(d.store, false, func(txn kv.Transaction) error {
+		err := d.verifySchemaMetaVersion(txn, is.SchemaMetaVersion())
```
Comment (shenli) [inline ddl/ddl.go]: If verify failed, why not get the latest info and continue to do ddl?

File: ddl/ddl.go [hunk 2]
```diff
@@ -138,16 +169,20 @@ func (d *ddl) DropSchema(ctx context.Context, schema model.CIStr) (err error) {
 
 	// Delete meta key
 	err = kv.RunInNewTxn(d.store, false, func(txn kv.Transaction) error {
+		err := d.verifySchemaMetaVersion(txn, is.SchemaMetaVersion())
+		if err != nil {
+			return errors.Trace(err)
```
Comment (shenli) [inline ddl/ddl.go]: Drop an un exists schema will not cause error if with IfNotExists clause.

File: ddl/ddl.go (hunk 1 above)
Comment (ngaut) [inline ddl/ddl.go]: It's complex. And administrator should be notified if schema changed.

File: ddl/ddl.go (hunk 2 above)
Comment (ngaut) [inline ddl/ddl.go]: We will refresh schema information every a few seconds.

**Comment by coocood** [comment]
LGTM

**Comment by ngaut** [comment]
PTAL @c4pt0r @siddontang

File: ddl/ddl.go (hunk 2 above)
Comment (shenli) [inline ddl/ddl.go]: ok

File: ddl/ddl.go [hunk 3]
```diff
@@ -622,25 +704,28 @@ func (d *ddl) DropIndex(ctx context.Context, schema, tableName, indexNmae model.
...
 		return errors.Trace(err)
 	}
-	err = kv.RunInNewTxn(d.store, false, func(txn kv.Transaction) error {
-		key := []byte(meta.DBMetaKey(info.ID))
-		if err := txn.LockKeys(key); err != nil {
-			return errors.Trace(err)
-		}
-		return txn.Set(key, b)
-	})
+	if err != nil {
```
Comment (shenli) [inline ddl/ddl.go]: err is checked twice

File: ddl/ddl.go [hunk 4]
```diff
@@ -82,27 +87,54 @@ func (d *ddl) CreateSchema(ctx context.Context, schema model.CIStr) (err error)
...
+	err = kv.RunInNewTxn(d.store, false, func(txn kv.Transaction) error {
+		err := d.verifySchemaMetaVersion(txn, is.SchemaMetaVersion())
+		if err != nil {
+			return errors.Trace(err)
+		}
+		err = d.writeSchemaInfo(info, txn)
+		return errors.Trace(err)
+	})
+	if d.onDDLChange != nil {
+		err = d.onDDLChange(err)
```
Comment (siddontang) [inline ddl/ddl.go]: here the ddl change err may overwrite above err if exists, is it ok?

File: ddl/ddl.go [hunk 5]
```diff
@@ -82,27 +87,54 @@ func (d *ddl) CreateSchema(ctx context.Context, schema model.CIStr) (err error)
...
+		return errors.Errorf("Schema changed, our version %d, got %d", schemaMetaVersion, curVer)
+	}
+
+	// Increment version
+	_, err = txn.Inc(meta.SchemaMetaVersionKey, 1)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	if err := txn.LockKeys(meta.SchemaMetaVersionKey); err != nil {
```
Comment (siddontang) [inline ddl/ddl.go]: err = txn.LockKeys() is ok
no need to check err here

File: ddl/ddl.go (hunk 4 above)
Comment (ngaut) [inline ddl/ddl.go]: Yes, we need onDDLChange to overwrite error.

**Comment by siddontang** [comment]
LGTM

**Pull request merged**